docs: add early-beta guide on using sso with saml (#12831)
supabase/supabase: apps/docs/pages/guides/auth/sso/auth-sso-saml.mdx

- **Emails are not necessarily unique.**  
  Given the behavior with no automatic linking, email addresses are no longer a unique identifier for a user account. Please always use the user's UUID to correctly reference user accounts.
- **Sessions may have a maximum duration.**  
  Depending on the configuration of the identity provider, a login session established with SSO may forcibly log out a user after a certain period of time.

### Row Level Security

You can use information about the SSO identity provider in Row Level Security policies.

Here are some commonly used statements to extract SSO related information from the user's JWT:

- `auth.jwt()#>>'{amr,0,method}'`
  Returns the name of the last method used to verify the identity of this user. With SAML SSO this is `sso/saml`.
- `auth.jwt()#>>'{amr,0,provider}'`
  Returns the UUID of the SSO identity provider used by the user to sign-in.
- `auth.jwt()#>>'{user_metadata,iss}'`
  Returns the identity provider's SAML 2.0 `EntityID`

<Admonition type="caution">
If you use [Multi-Factor Authentication](/guides/auth/auth-mfa) with SSO, the `amr` array may have a different method at index `0`!

> </Admonition>

Author: Right now you can't actually use MFA with SSO, GoTrue will reject the request, but this is only temporary and will be possible with general availability.

> A common use case with SSO is to use the UUID of the identity provider as the identifier for the organization the user belongs to. This way you can use restrictive RLS policies to scope down the SQL actions and data that a user is able to access based on what SSO provider was used to sign the user in.

Author: Mention "tenant" somewhere...

  -- the organization's SSO identity provider
  sso_provider_id uuid unique,
  -- name of the organization
  name text,
  -- billing plan (paid, free, enterprise)
  billing_plan text
);
```

You can use the information present in the user's JWT to scope down which rows from this table the user can see, without doing any additional user management:

```sql
create policy "View organization settings."
  on organization_settings
  as restrictive
  using (
    sso_provider_id = auth.jwt()#>>'{amr,0,provider}'
  );
```

## Managing SAML 2.0 connections

### Prerequisites

SSO support with SAML 2.0 is in an early beta release. This guide uses the following software which you need to install on your machine to configure your project:

- [**cURL**](https://curl.se)  
  It is typically pre-installed in macOS and GNU/Linux distributions.
- [**jq**](https://stedolan.github.io/jq/)  
  You can install it with `brew install jq` on macOS or using your distribution's package manager.

You would need access to two keys -- the `anon` and `service_role` key. You can obtain these on the [Project API Keys](https://app.supabase.com/project/_/settings/api) page in the dashboard.

> We publish an [OpenAPI specification](https://github.com/supabase/gotrue/blob/master/openapi.yaml) which you can refer to at any time.

Author: Obviously all of this will be replaced with well designed `supabase` CLI commands.

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  id: 'auth-sso-saml',
  title: 'Single Sign-On with SAML 2.0',
  description: 'Use Single Sign-On (SSO) authentication with SAML 2.0',
  video: 'https://www.youtube.com/v/em1cpOAXknM',
}

> Supabase Auth supports enterprise level Single Sign-On (SSO) using the SAML 2.0 protocol. In this document you will learn how you can manage connections to compatible identity providers.

Reviewer: ```suggestion
Supabase Auth supports enterprise-level Single Sign-On (SSO) for any Provides compatible with the using the SAML 2.0 protocol.
```

Reviewer: not sure exactly about this wording, but generally I like to avoid saying "what you will learn" - just get straight into it

<Admonition type="info">
This is an early beta release of these APIs. We are still working on adding support and a great developer experience in the CLI and Dashboard.

If you are comfortable using these APIs and would like to try out SSO with SAML 2.0 for your project, please [open a support ticket](https://app.supabase.com/support/new).

> We do not believe these APIs will change significantly until the feature is generally available, but we do reserve the right to modify them. Projects in the beta will be notified of any changes.

Reviewer: ```suggestion
These APIs are not expected to change before the feature is generally available, but we do reserve the right to modify them. Projects in the Beta will be notified of any changes.
```

</Admonition>

## Terminology

> When working with SAML and SSO, it often can feel like there's too many acronyms to keep track of. Here's what all of them mean.

Reviewer: ```suggestion
The number of SAML and SSO acronyms can often overwhelming. Here's a glossary which you can refer back to at any time:
```

- **Identity Provider**, **IdP** or **IDP**  

>   This is software that manages user accounts at a company or organization. It is able to verify the identity of a user and exchange that information with your Supabase project. Commonly used identity providers are: Microsoft Active Directory (Azure AD, Microsoft Entra), Okta, Google Workspaces (GSuite), PingIdentity, OneLogin, and many others.

Reviewer 2: ```suggestion
  This is software that manages user accounts at a company or organization. It can verify the identity of a user and exchange that information with your Supabase project. Commonly used identity providers are: Microsoft Active Directory (Azure AD, Microsoft Entra), Okta, Google Workspaces (GSuite), PingIdentity, OneLogin, and many others.
```

/nit

- **Service Provider**, **SP**  
  This is the software that is asking for user information from an identity provider. In Supabase, this is your project's Auth server.
- **Assertion**  
  An assertion is a statement issued by an identity provider that contains information about a user.
- **EntityID**  
  A globally unique ID (usually a URL) that identifies an Identity Provider or Service Provider across the world.
- **NameID**  
  A unique ID (usually an email address) that identifies a user at an Identity Provider.
- **Metadata**  
  An XML document that describes the features and configuration of an Identity Provider or Service Provider. It can be as a standalone document or as a URL. Usually (but not always) the `EntityID` is the URL at which you can access the Metadata.
- **Certificate**  
  Supabase Auth (the Service Provider) trusts assertions from an Identity Provider based on the signature attached to the assertion. The signature is verified according to the certificate present in the Metadata.

> - **ACS URL**  

Reviewer 3: ```suggestion
- **Assertion Consumer Service (ACS) URL**  
```

Reviewer 3: we should also add one for Single Logout (SLO):

```
- **Single Logout (SLO)**
Single Logout (SLO) allows a user to terminate all authentication sessions from other SPs that implement SLO by performing a single logout action.
```

  This is one of the most important SAML URLs. It is the URL where Supabase Auth will accept assertions from an identity provider. Basically, once the identity provider verifies the user's identity it will redirect to this URL and the redirect request will contain the assertion.

> - **Binding (Redirect, POST or Artifact)**  

Reviewer 2: ```suggestion
- **Binding (Redirect, POST, or Artifact)**  
```

/nit if we're going with oxford commas

  This is a description of the way an identity provider communicates with Supabase Auth. When using the Redirect binding, the communication occurs using HTTP 301 redirects. When it's `POST`, it's using `POST` requests sent with `<form>` elements on a page. When using Artifact, it's using a more secure exchange over a Redirect or `POST`.
- **RelayState**  
  State used by Supabase Auth to hold information about a request to verify the identity of a user.

## Important SAML 2.0 information

Below is information about your project's SAML 2.0 configuration which you can share with the company or organization that you're trying to on-board.

| Name         | Value                                                     |
| ------------ | --------------------------------------------------------- |
| EntityID     | `https://<project>.supabase.co/auth/v1/sso/saml/metadata` |
| Metadata URL | `https://<project>.supabase.co/auth/v1/sso/saml/metadata` |
| ACS URL      | `https://<project>.supabase.co/auth/v1/sso/saml/acs`      |
| SLO URL      | `https://<project>.supabase.co/auth/v1/sso/slo`           |
| NameID       | Required `emailAddress` or `persistent`                   |

> Note that SLO (Single Logout) is not supported at this time with Supabase Auth as it is a rarely supported feature by identity providers. However the URL is registered and advertised for when this does become available.

Reviewer 2: ```suggestion
Note that SLO (Single Logout) is not supported at this time with Supabase Auth as it is a rarely supported feature by identity providers. However, the URL is registered and advertised for when this does become available.
```

/nit consider adding comma

Append `?download=true` to the Metadata URL to get a downloadable Metadata XML file.

### User accounts and identities

User accounts and identities created via SSO differ from regular (email, phone, password, social login...) accounts in these ways:

- **No automatic linking.**  
  Each user account verified using a SSO identity provider will not be automatically linked to existing user accounts in the system. That is, if a user `jane.doe@company.com` had signed up with a password, and then uses their company SSO login with your project, there will be two `jane.doe@company.com` user accounts in the system.
- **Emails are not necessarily unique.**  

>   Given the behavior with no automatic linking, email addresses are no longer a unique identifier for a user account. Please always use the user's UUID to correctly reference user accounts.

Reviewer 2: Good shout. In the future, we can also consider highlighting this in other Auth sections too, perhaps in the [Email login page] or similar (https://supabase.com/docs/guides/auth/auth-email). Imagine that quite a few Auth users might assume email uniqueness but won't be using SSO.

})
```

Which will start the sign-in process using the SSO Identity Provider registered for the `company.com` domain name. If the SSO Identity Provider does not have an associated domain name, you can use `providerId` instead.

### Understanding attribute mappings

When a user signs in using the SAML 2.0 Single Sign-On protocol, an XML document called the SAML Assertion is exchanged between the Identity Provider and Supabase Auth.

This assertion contains information about the user's identity and other authentication information, such as:

- Unique ID of the user (called `NameID` in SAML)
- Email address
- Name of the user
- Department or organization
- Other attributes present in the users directory managed by the identity provider

Other than the unique ID of the user, SAML does not make it mandatory that any other attributes appear in the assertion. Identity Providers are configured about what user information is shared with your project.

Your project can be configured to recognize these attributes and map them into your project's database using a JSON structure. This process is called attribute mapping, and varies according to the configuration of the Identity Provider.

> For example the following JSON structure configures attribute mapping for the `email` and `first_name` user identity properties.

Reviewer 2: ```suggestion
For example, the following JSON structure configures attribute mapping for the `email` and `first_name` user identity properties.
```

/nit consider adding a comma
